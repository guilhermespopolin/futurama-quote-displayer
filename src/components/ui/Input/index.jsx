import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const errorModifier = ({ error, theme }) => {
  if (error) {
    return `
      border: 2px solid ${theme.colors.error};

      &:focus, &:active {
        border: 2px solid ${theme.colors.error};
      }
    `
  }

  return null
}

const InputWrapper = styled.div`
  display: inline-flex;
  flex-flow: column nowrap;
  padding: 0.5rem;
`

const StyledInput = styled.input`
  outline: none;
  border: none;
  padding: 0.5rem 0.3rem;
  background-color: transparent;
  border: 2px solid ${({ theme }) => theme.colors.complement4};
  font-family: 'Oswald', sans-serif;
  font-size: 1.1rem;
  color: var(--darkGray);

  &:focus, &:active {
    border: 2px solid ${({ theme }) => theme.colors.complement0}
  }

  ${errorModifier}
`

const InputError = styled.span`
  display: inline-block;
  height: 16px;
  margin-top: 0.5rem;
  font-family: 'Oswald', sans-serif;
  color: ${({ theme }) => theme.colors.error};
`

function Input({
  className,
  value,
  onChange,
  error,
  ...others
}) {
  return (
    <InputWrapper>
      <StyledInput
        value={value}
        onChange={onChange}
        error={error}
        {...others}
      />
      <InputError data-testid="input-error-message">{error}</InputError>
    </InputWrapper>
  )
}

Input.propTypes = {
  className: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func,
  error: PropTypes.string,
}

Input.defaultProps = {
  className: '',
  value: '',
  onChange: () => {},
  error: '',
}

export default Input
