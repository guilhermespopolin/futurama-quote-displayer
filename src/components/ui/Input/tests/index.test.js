import React from 'react'
import { render } from '__tests__/utils/rtlWrapper'

import Input from '..'

describe('<Input />', () => {
  it('should render properly', () => {
    const { getByPlaceholderText } = render(<Input placeholder="Enter some text" />)
    const inputNode = getByPlaceholderText('Enter some text')

    expect(inputNode).toMatchSnapshot()
  })

  describe('When there is some error', () => {
    it('should render properly', () => {
      const { getByPlaceholderText } = render(
        <Input
          placeholder="Enter some text"
          error="Something went wrong"
        />,
      )
      const inputNode = getByPlaceholderText('Enter some text')

      expect(inputNode).toMatchSnapshot()
    })
  })
})

