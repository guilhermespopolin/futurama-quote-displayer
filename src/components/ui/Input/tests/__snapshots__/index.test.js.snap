// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> When there is some error should render properly 1`] = `
.c0 {
  outline: none;
  border: none;
  padding: 0.5rem 0.3rem;
  background-color: transparent;
  border: 2px solid #b95900;
  font-family: 'Oswald',sans-serif;
  font-size: 1.1rem;
  color: var(--darkGray);
  border: 2px solid #d30d0d;
}

.c0:focus,
.c0:active {
  border: 2px solid #ff7a00;
}

.c0:focus,
.c0:active {
  border: 2px solid #d30d0d;
}

<input
  class="c0"
  placeholder="Enter some text"
  value=""
/>
`;

exports[`<Input /> should render properly 1`] = `
.c0 {
  outline: none;
  border: none;
  padding: 0.5rem 0.3rem;
  background-color: transparent;
  border: 2px solid #b95900;
  font-family: 'Oswald',sans-serif;
  font-size: 1.1rem;
  color: var(--darkGray);
}

.c0:focus,
.c0:active {
  border: 2px solid #ff7a00;
}

<input
  class="c0"
  placeholder="Enter some text"
  value=""
/>
`;
